src/renderer/api/kube-watch-api.ts

```diff
@@ -1,201 +1,326 @@
+  subscribeApi(api: KubeApi | KubeApi[]) {
+    const apis: KubeApi[] = [api].flat();
+
     apis.forEach(api => {
       this.subscribers.set(api, this.getSubscribersCount(api) + 1);
     });
 
-    return () => apis.forEach(api => {
-      const count = this.getSubscribersCount(api) - 1;
+    return () => {
+      apis.forEach(api => {
+        const count = this.getSubscribersCount(api) - 1;
 
-      if (count <= 0) this.subscribers.delete(api);
-      else this.subscribers.set(api, count);
-    });
+        if (count <= 0) this.subscribers.delete(api);
+        else this.subscribers.set(api, count);
+      });
+    };
   }
 
-  // FIXME: use POST to send apis for subscribing (list could be huge)
-  // TODO: try to use normal fetch res.body stream to consume watch-api updates
-  // https://github.com/lensapp/lens/issues/1898
-  protected async getQuery() {
-    const { namespaceStore } = await import("../components/+namespaces/namespace.store");
+  subscribeStores(stores: KubeObjectStore[], options: IKubeWatchSubscribeStoreOptions = {}): () => void {
+    const { preload = true, waitUntilLoaded = true } = options;
+    const loading: Promise<any>[] = [];
+    const disposers: Function[] = [];
+    let isDisposed = false;
 
-    await namespaceStore.whenReady;
-    const { isAdmin } = getHostedCluster();
+    async function subscribe() {
+      if (isDisposed) return;
+      const unsubscribeList = await Promise.all(stores.map(store => store.subscribe()));
 
-    return {
-      api: this.activeApis.map(api => {
-        if (isAdmin && !api.isNamespaced) {
-          return api.getWatchUrl();
-        }
+      disposers.push(...unsubscribeList);
+      if (isDisposed) unsubscribe();
+    }
 
-        if (api.isNamespaced) {
-          return namespaceStore.getContextNamespaces().map(namespace => api.getWatchUrl(namespace));
-        }
+    function unsubscribe() {
+      isDisposed = true;
+      disposers.forEach(dispose => dispose());
+      disposers.length = 0;
+    }
 
-        return [];
-      }).flat()
-    };
+    if (preload) {
+      loading.push(...stores.map(store => store.loadAll()));
+    }
+
+    if (waitUntilLoaded) {
+      Promise.all(loading).then(subscribe, error => {
+        this.log({
+          message: new Error("Loading stores has failed"),
+          meta: { stores, error, options },
+        });
+      });
+    } else {
+      subscribe();
+    }
+
+    return unsubscribe;
   }
 
-  // todo: maybe switch to websocket to avoid often reconnects
-  @autobind()
-  protected async connect() {
-    if (this.evtSource) this.disconnect(); // close previous connection
+  protected connectionCheck() {
+    this.log({
+      message: "connection check",
+      meta: { connected: this.isConnected },
+    });
 
-    const query = await this.getQuery();
+    if (this.isConnected) return;
+
+    return this.connect();
+  }
+
+  protected async connect(apis = this.apis) {
+    this.disconnect(); // close active connections first
```

Comment:
Name is expected. Disconnect is happening here because we keep only single stream-connection and it comes mostly from `mobx`-reaction that triggered when observable watch-apis changed. 
Disconnect itself also used separately in other place.